File: tools/release/check_publish_environment.py
#!/usr/bin/env python3
"""Fail fast when selected release products are missing publish credentials."""

from __future__ import annotations

import argparse
import json
import os
import sys
from typing import NoReturn

import product_metadata

OIDC_TARGETS = {"crates-io", "npm", "jsr"}
MAVEN_TARGETS = {"maven-central"}
GITHUB_TARGETS = {"github-release", "github-release-assets", "swift-package-source-tag"}
FORBIDDEN_ENV_VARS = {
    "CARGO_REGISTRY_TOKEN": (
        {"crates-io"},
        "Cargo publishing uses crates.io trusted publishing through GitHub Actions OIDC",
    ),
    "NPM_TOKEN": (
        {"npm"},
        "npm publishing uses trusted publishing with provenance through GitHub Actions OIDC",
    ),
    "NODE_AUTH_TOKEN": (
        {"npm"},
        "npm publishing uses trusted publishing with provenance through GitHub Actions OIDC",
    ),
    "JSR_TOKEN": ({"jsr"}, "JSR publishing uses GitHub Actions OIDC"),
    "COCOAPODS_TRUNK_TOKEN": (
        set(),
        "Apple SDK releases use SwiftPM plus GitHub assets, not CocoaPods trunk",
    ),
    "COCOAPODS_TRUNK_EMAIL": (
        set(),
        "Apple SDK releases use SwiftPM plus GitHub assets, not CocoaPods trunk",
    ),
}


def fail(message: str) -> NoReturn:
    print(f"check_publish_environment.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def parse_products(raw: str) -> set[str]:
    value = json.loads(raw)
    if not isinstance(value, list) or not all(isinstance(item, str) for item in value):
        fail("--products-json must be a JSON string list")
    products = set(value)
    known = set(product_metadata.product_ids())
    unknown = sorted(products - known)
    if unknown:
        fail(f"unknown release products: {', '.join(unknown)}")
    return products


def require_env(name: str, context: str, failures: list[str]) -> None:
    if not os.environ.get(name):
        failures.append(f"{context} requires {name}")


def require_any_env(names: list[str], context: str, failures: list[str]) -> None:
    if not any(os.environ.get(name) for name in names):
        failures.append(f"{context} requires one of {', '.join(names)}")


def selected_publish_targets(products: set[str]) -> set[str]:
    targets: set[str] = set()
    graph = product_metadata.load_graph()
    for product in products:
        config = product_metadata.product_config(product, graph)
        targets.update(product_metadata.string_list(config, "publish_targets", product))
    return targets


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--products-json", required=True)
    args = parser.parse_args(argv)

    products = parse_products(args.products_json)
    publish_targets = selected_publish_targets(products)
    failures: list[str] = []

    for name, (blocked_targets, reason) in sorted(FORBIDDEN_ENV_VARS.items()):
        applies_to_selection = bool(products) and (
            not blocked_targets or bool(publish_targets & blocked_targets)
        )
        if applies_to_selection and os.environ.get(name):
            failures.append(f"forbidden release credential {name} is set: {reason}")

    if publish_targets & OIDC_TARGETS:
        require_env("ACTIONS_ID_TOKEN_REQUEST_TOKEN", "trusted publishing", failures)
        require_env("ACTIONS_ID_TOKEN_REQUEST_URL", "trusted publishing", failures)

    if publish_targets & GITHUB_TARGETS:
        require_any_env(["GH_TOKEN", "GITHUB_TOKEN"], "GitHub release assets and tags", failures)

    if publish_targets & MAVEN_TARGETS:
        for name in [
            "ORG_GRADLE_PROJECT_mavenCentralUsername",
            "ORG_GRADLE_PROJECT_mavenCentralPassword",
            "ORG_GRADLE_PROJECT_signingInMemoryKey",
            "ORG_GRADLE_PROJECT_signingInMemoryKeyId",
            "ORG_GRADLE_PROJECT_signingInMemoryKeyPassword",
        ]:
            require_env(name, "Maven Central publish", failures)

    if failures:
        fail("missing publish environment:\n  - " + "\n  - ".join(failures))

    print("publish environment checks passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: tools/release/check_publish_environment.mjs
#!/usr/bin/env bun
import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
const oidcTargets = new Set(['crates-io', 'npm', 'jsr']);
const mavenTargets = new Set(['maven-central']);
const githubTargets = new Set(['github-release', 'github-release-assets', 'swift-package-source-tag']);
const forbiddenEnvVars = {
  CARGO_REGISTRY_TOKEN: [
    new Set(['crates-io']),
    'Cargo publishing uses crates.io trusted publishing through GitHub Actions OIDC',
  ],
  NPM_TOKEN: [
    new Set(['npm']),
    'npm publishing uses trusted publishing with provenance through GitHub Actions OIDC',
  ],
  NODE_AUTH_TOKEN: [
    new Set(['npm']),
    'npm publishing uses trusted publishing with provenance through GitHub Actions OIDC',
  ],
  JSR_TOKEN: [new Set(['jsr']), 'JSR publishing uses GitHub Actions OIDC'],
  COCOAPODS_TRUNK_TOKEN: [
    new Set(),
    'Apple SDK releases use SwiftPM plus GitHub assets, not CocoaPods trunk',
  ],
  COCOAPODS_TRUNK_EMAIL: [
    new Set(),
    'Apple SDK releases use SwiftPM plus GitHub assets, not CocoaPods trunk',
  ],
};

function fail(message) {
  console.error(`check_publish_environment.mjs: ${message}`);
  process.exit(1);
}

function parseArgs(argv) {
  let productsJson = null;
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === '--products-json') {
      productsJson = argv[index + 1] ?? null;
      index += 1;
      continue;
    }
    fail(`unknown argument: ${arg}`);
  }
  if (productsJson === null) {
    fail('usage: tools/release/check_publish_environment.mjs --products-json <json-string-list>');
  }
  return { productsJson };
}

function parseProducts(raw) {
  let value;
  try {
    value = JSON.parse(raw);
  } catch (error) {
    fail(`--products-json must be valid JSON: ${error.message}`);
  }
  if (!Array.isArray(value) || value.some((item) => typeof item !== 'string')) {
    fail('--products-json must be a JSON string list');
  }
  return new Set(value);
}

async function productConfigs() {
  const releasePlease = JSON.parse(await fs.readFile(path.join(root, 'release-please-config.json'), 'utf8'));
  if (typeof releasePlease.packages !== 'object' || releasePlease.packages === null) {
    fail('release-please-config.json must define packages');
  }
  const products = new Map();
  const packageEntries = Object.entries(releasePlease.packages).sort(([left], [right]) =>
    left < right ? -1 : left > right ? 1 : 0,
  );
  for (const [packagePath, packageConfig] of packageEntries) {
    if (path.isAbsolute(packagePath) || packagePath.split(/[\\/]/u).includes('..')) {
      fail(`release-please package path must stay inside the repository: ${packagePath}`);
    }
    const component = packageConfig?.component;
    if (typeof component !== 'string' || component.length === 0) {
      fail(`${packagePath}.component must be a non-empty string`);
    }
    const file = path.join(root, packagePath, 'release.toml');
    const metadata = Bun.TOML.parse(await fs.readFile(file, 'utf8'));
    const id = metadata.id;
    if (id !== component) {
      fail(`${path.relative(root, file)} must declare id = "${component}"`);
    }
    if (products.has(id)) {
      fail(`duplicate release product id ${id}`);
    }
    const publishTargets = metadata.publish_targets ?? [];
    if (
      !Array.isArray(publishTargets) ||
      publishTargets.some((target) => typeof target !== 'string')
    ) {
      fail(`${id}.publish_targets must be a string list`);
    }
    products.set(id, { publishTargets });
  }
  return products;
}

function requireEnv(name, context, failures) {
  if (!process.env[name]) {
    failures.push(`${context} requires ${name}`);
  }
}

function requireAnyEnv(names, context, failures) {
  if (!names.some((name) => process.env[name])) {
    failures.push(`${context} requires one of ${names.join(', ')}`);
  }
}

function intersects(left, right) {
  for (const value of left) {
    if (right.has(value)) {
      return true;
    }
  }
  return false;
}

const args = parseArgs(Bun.argv.slice(2));
const products = parseProducts(args.productsJson);
const configs = await productConfigs();
const unknown = [...products].filter((product) => !configs.has(product)).sort();
if (unknown.length > 0) {
  fail(`unknown release products: ${unknown.join(', ')}`);
}

const publishTargets = new Set();
for (const product of products) {
  for (const target of configs.get(product).publishTargets) {
    publishTargets.add(target);
  }
}

const failures = [];
for (const [name, [blockedTargets, reason]] of Object.entries(forbiddenEnvVars).sort()) {
  const appliesToSelection =
    products.size > 0 && (blockedTargets.size === 0 || intersects(publishTargets, blockedTargets));
  if (appliesToSelection && process.env[name]) {
    failures.push(`forbidden release credential ${name} is set: ${reason}`);
  }
}

if (intersects(publishTargets, oidcTargets)) {
  requireEnv('ACTIONS_ID_TOKEN_REQUEST_TOKEN', 'trusted publishing', failures);
  requireEnv('ACTIONS_ID_TOKEN_REQUEST_URL', 'trusted publishing', failures);
}

if (intersects(publishTargets, githubTargets)) {
  requireAnyEnv(['GH_TOKEN', 'GITHUB_TOKEN'], 'GitHub release assets and tags', failures);
}

if (intersects(publishTargets, mavenTargets)) {
  for (const name of [
    'ORG_GRADLE_PROJECT_mavenCentralUsername',
    'ORG_GRADLE_PROJECT_mavenCentralPassword',
    'ORG_GRADLE_PROJECT_signingInMemoryKey',
    'ORG_GRADLE_PROJECT_signingInMemoryKeyId',
    'ORG_GRADLE_PROJECT_signingInMemoryKeyPassword',
  ]) {
    requireEnv(name, 'Maven Central publish', failures);
  }
}

if (failures.length > 0) {
  fail(`missing publish environment:\n  - ${failures.join('\n  - ')}`);
}

console.log('publish environment checks passed');
